### ISSUE TITLE
Add API to get RST_STREAM value from HTTP exceptions

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#43239]
Part of User Story: https://github.com/dotnet/core/issues/5493

## Background and Motivation

gRPC specifies that RST_STREAM error codes are mapped to gRPC error codes - https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#errors

I believe that today the RST_STREAM value is only exposed in exception message descriptions. The only way to get this value is to parse exception messages, which is pretty awful for the usual reasons.

There should be an exception property with the RST_STREAM error code so it can be used programmatically.

## Proposed API

Make this value publicly accessible:

https://github.com/dotnet/runtime/blob/faf91e0a7495b70bdae6a8e7109b612a47a74662/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2ProtocolException.cs#L28

I don't feel strongly that it needs to be an enum. `public int ProtocolError { get; }` would also be fine.

**Comment by geoffkizer** [comment]
I agree that we should expose this value.

There's a similar issue with error codes from GOAWAY -- this is similar to RST_STREAM, except it affects the whole connection and kills all streams. As part of exposing the RST_STREAM error code, we should also consider exposing the GOAWAY error code if/when appropriate.

We should also try to do this in such a way that it works for HTTP3 as well, since there is a similar error code capability in HTTP3. Not sure to what extent the error codes are aligned, though...

**Comment by JamesNK** [comment]
HTTP/3 values are distinct https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-11.2.3

**Comment by geoffkizer** [comment]
> HTTP/3 error codes are distinct

Of course they are :( 

And they can be 62 bits long, too.

I guess this argues for having something like `long ProtocolError`, with values that are version specific. But that raises the issue of how you know which protocol version was actually used...

**Comment by JamesNK** [comment]
It looks like distinct ranges for each protocol.

Http2: <255
Http3: >256

**Comment by karelz** [comment]
As discussed earlier, we won't be able to make this in 6.0. Moving to Future.

**Comment by JamesNK** [comment]
Behold the hack I added to gRPC: https://github.com/JamesNK/grpc-dotnet/blob/db3e63c12188e865481f952e81402d6ec64c37cc/src/Grpc.Net.Client/Internal/GrpcProtocolHelpers.cs#L422-L561 😄

**Comment by antonfirsov** [comment]
I also agree that we should create a unified proposal that covers RST_STREAM, GOAWAY, and the HTTP/3 cases. What makes this tricky is the way the exception/error information is nested today:

- Http2: `HttpRequestException --> IOException  --> Http2StreamException/Http2ConnectionException --> ProtocolError`
- Http3 : `HttpRequestException --> QuicStreamAbortedException --> ErrorCode` (At least in cases I was able to figure out.)

One possible solution is to expose the exception types `Http2StreamException`, `Http2ConnectionException` and `QuicStreamAbortedException` and leave it for the user to dispatch the nested error information. If we go down this way, I wonder if the way we nest the exceptions today is an intentional behavioral contract, or is it something we can change it in order to simplify things and make sure `HttpRequestException.InnerException` is an instance of the 3 possible "protocol exceptions" in case they are present?

If simplification is not possible, we can create a shortcut as a separate property on `HttpRequestException`:

```C#
public class HttpRequestException {
    // Http2StreamException, Http2ConnectionException or QuicStreamAbortedException:
    public Exception? ProtocolException { get; } 
}
```

A different approach would be to expose protocol error code right on `HttpRequestException`, however the gRPC workaround code in https://github.com/dotnet/runtime/issues/43239#issuecomment-934124102 indicates that we can't  get away with a single `long ErrorCode` property, because gRPC needs to distinguish between stream and connection errors, only mapping the stream errors. @JamesNK correct me if I'm wrong on this.

A possible API:

```C#
public class HttpRequestException {
    long? ProtocolErrorCode { get; }
    ProtocolErrorSource? ProtocolErrorSource { get; }
}

public ProtocolErrorSource {
    Http2Stream,
    Http2Connection,
    QuicStream
}
```

@dotnet/ncl thoughts?

**Comment by JamesNK** [comment]
> however the gRPC workaround in https://github.com/dotnet/runtime/issues/43239#issuecomment-934124102 indicates that we can't get away with a single long ErrorCode property, because gRPC needs to distinguish between stream and connection errors, only mapping the stream errors. @JamesNK correct me if I'm wrong on this.

I don't think we need a way to distinguish.

When I experimented with how to get error codes, which resulted in the exception message parsing code I linked to, all the thrown errors I found were either Http2StreamException or QuicStreamAbortedException. I didn't find any examples where I cared about Http2ConnectionException. I'm guessing because I cared about catching errors when a request is in-progress

When is Http2ConnectionException thrown?

tldr `long? ProtocolErrorCode { get; }` by itself is probably fine.

**Comment by antonfirsov** [comment]
> When is `Http2ConnectionException` thrown?

1. Whenever we receive a GOAWAY frame
2. Whenever we encounter a protocol violation that requires us to terminate the connection

Since your `ResolveRpcExceptionStatusCode` method is currently filtering on `Http2StreamException` only, I assumed you want to avoid mapping protocol error codes when these events occur, but maybe I was wrong?

**Comment by JamesNK** [comment]
I care about those as well. They never showed up in my tests so I never needed to check for Http2ConnectionException.

**Comment by JamesNK** [comment]
@karelz This was moved from 6.0 to future last year, but it was never added to the 7.0 milestone. Should it be?

**Comment by antonfirsov** [comment]
@JamesNK here it is in 7.0 😄 

@dotnet/ncl can we think of cases when a user would need to see which of the many possible scenarios has resulted in `HttpRequestException` reporting a protocol error, or are we fine simply adding a field with the error code?

**Comment by karelz** [comment]
@JamesNK sorry, that was an oversight on our side. Thanks @antonfirsov for fixing it!

**Comment by JamesNK** [comment]
> @JamesNK here it is in 7.0 😄
> 
> @dotnet/ncl can we think of cases when a user would need to see which of the many possible scenarios has resulted in `HttpRequestException` reporting a protocol error, or are we fine simply adding a field with the error code?

Are there places where a protocol error occurred but HttpRequestExveption isn’t the exception thrown?

**Comment by JamesNK** [comment]
Here is at least one case where `IOException` is thrown because a protocol error: https://github.com/dotnet/runtime/issues/62228

In this situation, `HttpResponseMessage` has successfully returned from `SendAsync` and the error happens when reading from the response body.

**Comment by antonfirsov** [comment]
Well, this opens up a bunch of options:
- View #62228 as an entirely separate issue (maybe it's enough to report a more verbose exception message?)
- Subclass `IOException` adding an error code, which we should always embed as `HttpRequestException.InnerException`
-  Subclass `IOException` adding an error code, but also repeat the error code property on `HttpRequestException`

**Comment by antonfirsov** [comment]
Long story short, it looks like we can get away with a simple API addition for this particular issue:

```C#
public class HttpRequestException {
    long? ProtocolErrorCode { get; }
}
```

We can extend this with more error details later if we want. We can utilize it to solve #62228.

**Comment by antonfirsov** [comment]
Team discussion:
It might be actually better to carry this information in an `IOException` subclass, maybe the same one for both HTTP2 and QUIC.
Passing this over to @rzikm who is responsible for the QUIC exception API design. I can help with the implementation when API-s are approved.

**Comment by ManickaP** [comment]
This is also related to decision whether to derive `QuicException` from `IOException`: https://github.com/dotnet/runtime/issues/56954
In such case, we would have `HttpRequestException` with `IOException` as the inner exception in both H/2 and H/3 cases. From that point of view, it might make more sense to reuse the similarity and design a common vehicle for the protocol error code. Moreover, this would also be useful for https://github.com/dotnet/runtime/issues/62228 without sticking `HttpRequestException` inside `IOException` (reverse order to what `HttpClient` does).

This line of thinking depends on the decision we'll make in #56954. @rzikm, could you please take this into account when you're thinking about QUIC exceptions?

Either way, as a fallback we can always do what https://github.com/dotnet/runtime/issues/43239#issuecomment-1125131300 above ^ suggests.

**Comment by antonfirsov** [comment]
Closing this in favor of #70684
